packages/cardpay-reward-programs/cardpay_reward_programs/rules/staking.py

```diff
@@ -0,0 +1,107 @@
+                from {token_holder_table} as tht, filtered_safes as sot
+                where tht.safe = sot.safe
+                and sot.type = 'depot'
+            ),
+            balance_changes as (select
+                safe,
+                balance_int64 - lag(balance_int64, 1 ,0) over (partition by safe order by _block_number asc) as change,
+                $5::float+1 as interest_rate,
+                ($2::integer - _block_number::integer) / $4::float  as compounding_rate,
+                from filtered_balances
+                where _block_number::integer < $2::integer
+                and token = $3::text
+                and safe is not null
+                qualify
+                _block_number::integer >= $1::integer
+                ),
+            original_balances as (select safe,
+                last(balance_int64) as change,
+                $5::float+1 as interest_rate,
+                1 as compounding_rate,
+                from filtered_balances a
+                where _block_number::integer < $1::integer
+                and token = $3::text
+                and safe is not null
+                and _block_number = (
+                    SELECT MAX(_block_number)
+                    FROM filtered_balances b
+                    WHERE a.safe = b.safe
+                    AND _block_number::integer < $1::integer
+                )
+                group by safe
+                ),
+            all_data as (select * from original_balances union all select * from balance_changes)
+
+            select owner as payee, sum(change * ((interest_rate**compounding_rate) - 1)) as rewards
+            from all_data
+            left join filtered_safes using (safe)
+            group by payee
+        """
+
+    def run(self, payment_cycle: int, reward_program_id: str):
+        start_block, end_block = (
+            payment_cycle - self.payment_cycle_length,
+            payment_cycle,
+        )
+        vars = [
+            start_block,  # $1 -> int
+            end_block,  # $2 -> int
+            self.token,  # $3 -> str
+            self.payment_cycle_length,  # $4 -> int
+            self.interest_rate_monthly,  # $5 -> float
+        ]
+
+        table_query = self._get_table_query(
+            "token_holder", "token_holder", self.start_block, payment_cycle
+        )
+
+        aux_table_query = self._get_table_query(
+            "safe_owner", "safe_owner", self.start_block, payment_cycle
+        )
+
+        if table_query == "parquet_scan([])" or aux_table_query == "parquet_scan([])":
+            df = pd.DataFrame(columns=["payee", "rewards"])
+        else:
+            df = self.run_query(table_query, vars, aux_table_query)
+        df["rewardProgramID"] = reward_program_id
+        df["paymentCycle"] = payment_cycle
+        df["validFrom"] = payment_cycle
+        df["validTo"] = payment_cycle + self.duration
+        df["token"] = self.token
+        df["amount"] = df["rewards"] * 1_000_000_000
```

Comment:
The balances exported are scaled down by 1e9 to make them manageable but with still pretty high precision (9 decimal places) and so the rewards are scaled up at the end here.